Bug 1278981: Add service to sign Stub Attribution URL params

pmac opened the pull request:
Uses HMAC with SHA256

pmac commented:
@jgmize this is ready for review. There are still some open questions, but this is the correct implementation based on the current situation.

/cc @ckprice

alexgibson commented:
I gave this a test locally by making an ajax GET request along with some params and it seems to work well 👍 

I'll defer to Josh on the code review side of things, but functionally I think this looks good.

stephendonner commented:
Hey @pmac @jgmize - I'd like to start being able to test this, coupled with https://github.com/mozilla-services/go-bouncer/pull/36 and, I think, the as-yet-to-be-coded/merged/stood-up Stub Attribution-download service.  Any chance of/when can we get this + the other PRs in, so I can test more fully?  Thanks!  /cc @oremj

jgmize commented:
I've only taken a cursory look at this code so far, given the current status (currently "Do Not Merge" + WIP), but I haven't seen anything obviously wrong yet. This is in need of a rebase, but that's not a blocker to deploying to a demo instance if needed.

alexgibson commented:
@stephendonner this is only one piece of the work that's required in order to test stub attribution. We're working on this along with the other parts this sprint. When things are ready for testing, we'll let you know. Thanks

stephendonner commented:
Thx @alexgibson - as you can tell/as usual, I'm anxious to test ;-)

pmac commented:
@jgmize it's ready for r? for real now.

@alexgibson I've made some changes. I changed it to expect exactly the params that should be in the URL of the page, so if you wanted to you could just tack it onto the AJAX request and let the service work it out. I also changed the returned JSON to use the parameter names that bouncer expects (`attribution_code` and `attribution_sig`).

alexgibson commented on bedrock/firefox/views.py:
```diff
@@ -68,6 +77,19 @@
+
+# available variations for onboarding first & second run tests - Q2 2016
+# https://bugzilla.mozilla.org/show_bug.cgi?id=1259608
+ONBOARDING_VARIATIONS = ['1', '2', '3', '4', '5', '6']
```
Is this needed here? rebase gremlins maybe? 👹

pmac replied:
Bah! Indeed. I'll remove.

pmac commented on bedrock/firefox/views.py:
```diff
@@ -86,6 +103,57 @@ def installer_help(request):
+        return HttpResponseJSON({'error': 'HMAC key not set'}, status=400)
+
+    rate = settings.STUB_ATTRIBUTION_RATE
+    if rate < 1 and random() > rate:
```
Hmm... I think you're right. Good catch.

pmac commented on bedrock/firefox/tests/test_views.py:
```diff
@@ -3,18 +3,160 @@
+        self.assertEqual(data['error'], 'no params')
+
+    def test_some_valid_param_data(self):
+        params = {'utm_source': 'brandt', 'utm_content': 'ae#her'}
```
Ah yes, good call. A better test would be a more likely invalid character. I'll fix.

pmac commented:
Updated to remove dynamic rate limiting and improve cache header usage.

pmac commented:
I think we're good here. Plus, this will not be used in prod until the frontend work is done, and it won't be turned on until we're ready and have an HMAC key. I'm merging this based on the previous feedback so that @alexgibson can more easily file his PR for the frontend bits.

Thanks all!
